internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// Create creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func Create(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan interface{}, 10)
+	jobMap[uintptr(jobHandle)] = iocpChan
+
+	defer func() {
+		if err != nil {
+			windows.Close(jobHandle)
+			delete(jobMap, uintptr(jobHandle))
+		}
+	}()
+
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		if err != nil {
+			initIOErr = err
+			return
+		}
+		ioCompletionPort = h
+		go pollIOCP(h)
+	})
+
+	if initIOErr != nil {
+		return nil, err
+	}
+
+	if err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+		return nil, err
+	}
+
+	return &JobObject{
+		jobHandle,
+		iocpChan,
+		sync.RWMutex{},
+	}, nil
+}
+
+// NotificationChan returns a channel that can be used to receive notifications surrounding
+// the job object. The caller will have to type assert the type of message received.
+func (job *JobObject) NotificationChan() chan interface{} {
+	return job.notifChan
+}
+
+// Close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *JobObject) Close() error {
+	if job.jobHandle == 0 {
+		return nil
+	}
+	job.handleLock.Lock()
+	defer job.handleLock.Unlock()
+
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			return err
+		}
+		// Handles now invalid so if the map entry to receive notifications for this job still
+		// exists remove it so we can stop receiving notifications.
+		if _, ok := jobMap[uintptr(job.jobHandle)]; ok {
+			delete(jobMap, uintptr(job.jobHandle))
+		}
+		job.jobHandle = 0
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *JobObject) Assign(pid uint32) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if pid == 0 {
+		return errors.New("process has not started")
```

Comment:
This is kind of a weird error message. I would expect something more like `invalid pid: 0`.

Reply:
I forget if this was for some weird behavior in exec.Cmd but regardless it shouldn't be tied to that behavior. Will change